### PR TITLE
Fixes for zero-sized allocation

### PR DESCRIPTION
**Pull request by isaachier**
See #1044.

### CODE REVIEW

File: src/analyze.cpp [hunk 1]
```diff
@@ -1848,10 +1849,10 @@ static void resolve_struct_type(CodeGen *g, TypeTableEntry *struct_type) {
     }
 
     assert(!struct_type->data.structure.zero_bits_loop_flag);
-    assert(struct_type->data.structure.fields);
```
Comment (isaachier) [inline src/analyze.cpp]: Failed assertion.

File: src/analyze.cpp [hunk 2]
```diff
@@ -1848,10 +1849,10 @@ static void resolve_struct_type(CodeGen *g, TypeTableEntry *struct_type) {
...
     assert(decl_node->type == NodeTypeContainerDecl);
 
     size_t field_count = struct_type->data.structure.src_field_count;
+    assert(struct_type->data.structure.fields != nullptr || field_count == 0);
```
Comment (isaachier) [inline src/analyze.cpp]: Attempted tweak.

File: src/analyze.cpp [hunk 3]
```diff
@@ -1964,7 +1965,7 @@ static void resolve_struct_type(CodeGen *g, TypeTableEntry *struct_type) {
         struct_type->di_type = replacement_di_type;
         return;
     }
-    assert(struct_type->di_type);
```
Comment (isaachier) [inline src/analyze.cpp]: Failed assertion.

File: src/analyze.cpp [hunk 4]
```diff
@@ -1964,7 +1965,7 @@ static void resolve_struct_type(CodeGen *g, TypeTableEntry *struct_type) {
...
+    assert(struct_type->di_type != nullptr || field_count == 0);
```
Comment (isaachier) [inline src/analyze.cpp]: Attempted tweak.

File: src/analyze.cpp [hunk 5]
```diff
@@ -1976,7 +1977,8 @@ static void resolve_struct_type(CodeGen *g, TypeTableEntry *struct_type) {
     // if you hit this assert then probably this type or a related type didn't
     // get ensure_complete_type called on it before using it with something that
     // requires a complete type
-    assert(LLVMStoreSizeOfType(g->target_data_ref, struct_type->type_ref) > 0);
```
Comment (isaachier) [inline src/analyze.cpp]: Failed assertion.

File: src/analyze.cpp [hunk 6]
```diff
@@ -1976,7 +1977,8 @@ static void resolve_struct_type(CodeGen *g, TypeTableEntry *struct_type) {
...
+    assert(LLVMStoreSizeOfType(g->target_data_ref, struct_type->type_ref) > 0 ||
```
Comment (isaachier) [inline src/analyze.cpp]: Attempted tweak.

File: src/codegen.cpp [hunk 7]
```diff
@@ -5922,20 +5922,20 @@ static const uint8_t int_sizes_in_bits[] = {
 };
 
 struct CIntTypeInfo {
```
Comment (isaachier) [inline src/codegen.cpp]: Padding fix discussed in #993.

File: src/util.hpp [hunk 8]
```diff
@@ -73,6 +76,9 @@ ATTRIBUTE_RETURNS_NOALIAS static inline T *allocate_nonzero(size_t count) {
 
 template<typename T>
 ATTRIBUTE_RETURNS_NOALIAS static inline T *allocate(size_t count) {
+    if (count == 0) {
```
Comment (isaachier) [inline src/util.hpp]: Makes behavior for zero sized allocation portable.

File: src/util.hpp [hunk 9]
```diff
@@ -92,21 +98,20 @@ static inline void safe_memcpy(T *dest, const T *src, size_t count) {
...
-    }
     return ptr;
 }
 
 template<typename T>
-static inline T *reallocate_nonzero(T *old, size_t old_count, size_t new_count) {
-    T *ptr = reinterpret_cast<T*>(realloc(old, new_count * sizeof(T)));
-    if (!ptr)
-        zig_panic("allocation failed");
+static inline T *reallocate(T *old, size_t old_count, size_t new_count) {
```
Comment (isaachier) [inline src/util.hpp]: Implement `reallocate` using `reallocate_nonzero` to avoid code duplication.

File: src/ir.cpp [hunk 10]
```diff
@@ -7647,7 +7647,7 @@ static TypeTableEntry *get_error_set_intersection(IrAnalyze *ira, TypeTableEntry
         assert(errors[error_entry->value] == nullptr);
         errors[error_entry->value] = error_entry;
     }
-    ZigList<ErrorTableEntry *> intersection_list = {};
+    ZigList<ErrorTableEntry *> intersection_list = {0};
```
Comment (isaachier) [inline src/ir.cpp]: You will see this in a few places. `ZigList` needs to be initialized in order to guarantee the members are zeroed.

Comment (andrewrk) [reply src/ir.cpp]: Can you cite a source showing that `{0}` is necessary and `{}` is not sufficient?

Comment (isaachier) [reply src/ir.cpp]: Actually you are right: https://stackoverflow.com/a/1069634/1930331. Maybe I'm thinking about C, but C++ will work.

**Pull request closed without merging**